fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/SyntheticRecordTypeBuilder.java

```diff
@@ -0,0 +1,186 @@
+     */
+    @Nonnull
+    public List<C> getConstituents() {
+        return constituents;
+    }
+
+    @Nonnull
+    protected abstract C newConstituent(@Nonnull String name, @Nonnull RecordTypeBuilder recordType);
+
+    /**
+     * Add a new constituent.
+     * @param constituent the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    protected C addConstituent(@Nonnull C constituent) {
+        constituents.add(constituent);
+        return constituent;
+    }
+
+    /**
+     * Add a new constituent by name.
+     * @param name the correlation name for the new constituent
+     * @param recordType the record type for the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    public C addConstituent(@Nonnull String name, @Nonnull RecordTypeBuilder recordType) {
+        return addConstituent(newConstituent(name, recordType));
+    }
+
+    /**
+     * Add a new constituent by name.
+     * @param name the correlation name for the new constituent
+     * @param recordType the name of the record type for the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    public C addConstituent(@Nonnull String name, @Nonnull String recordType) {
+        return addConstituent(name, metaDataBuilder.getRecordType(recordType));
+    }
+
+    /**
+     * Add a new constituent.
+     * @param recordType the record type for the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    public C addConstituent(@Nonnull RecordTypeBuilder recordType) {
+        return addConstituent(recordType.getName(), recordType);
+    }
+
+    /**
+     * Add a new constituent.
+     * @param constituent the correlation name and name of the record type for the new constituent
+     * @return the newly added constituent
+     */
+    @Nonnull
+    protected C addConstituent(@Nonnull String constituent) {
+        return addConstituent(constituent, constituent);
+    }
+
+    @Nonnull
+    @SuppressWarnings("squid:S1452")
+    public abstract SyntheticRecordType<?> build(@Nonnull RecordMetaData metaData, @Nonnull Descriptors.FileDescriptor fileDescriptor);
+
+    public void buildDescriptor(@Nonnull DescriptorProtos.FileDescriptorProto.Builder fileDescriptorProto) {
+        final DescriptorProtos.DescriptorProto.Builder descriptorProto = fileDescriptorProto.addMessageTypeBuilder();
+        descriptorProto.setName(name);
+        int fieldNumber = 0;
+        for (Constituent constituent : constituents) {
+            descriptorProto.addFieldBuilder()
+                    .setName(constituent.getName())
+                    .setNumber(++fieldNumber)
+                    .setType(DescriptorProtos.FieldDescriptorProto.Type.TYPE_MESSAGE)
+                    .setTypeName(constituent.getRecordType().getDescriptor().getFullName());
+        }
+    }
+
+    @Nonnull
+    protected KeyExpression buildPrimaryKey() {
```

Comment:
The primary key that gets produced here--can one evaluate a synthetic record against it and get something reasonable (like the two record primary keys back)?

Reply:
Okay, I think the answer to this question is "yes" now that it nests the primary key expression within a field named by the constituent name.

Reply:
There is now a test that some synthetic records with (fairly) complicated primary keys evaluate the same from the key in the meta-data as is in the synthesized record.